### PR TITLE
feat(clients/java): Optional read and connect timeout for an OAuth2 credentials client.

### LINKED ISSUE #3556: Add OAuth timeout config in Java client
**Is your feature request related to a problem? Please describe.**
The Go client allows the user to specify a timeout for the OAuth request through code and an environment variable.

**Describe the solution you'd like**
Allow user to configure a timeout for the OAuth request through code and an env var.

### PR DESCRIPTION
**Pull request by aivinog1**
## Description

I've added optional timeouts for a credentials client.

## Related issues

closes #3556

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [X] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [X] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

**Review by miguelpires — CHANGES_REQUESTED** [review]

**Comment by MiguelPires** [comment]
Thanks for the contribution @aivinog1! I added a few small comments. Please feel free to challenge anything you don't agree with.

File: clients/java/src/main/java/io/zeebe/client/impl/oauth/OAuthCredentialsProviderBuilder.java [hunk 1]
```diff
@@ -163,6 +218,24 @@ private void validate() {
         throw new IllegalArgumentException(
             "Expected specified credentials cache to be a file but found directory instead.");
       }
+      if (connectTimeout.isZero() || connectTimeout.isNegative()) {
+        throw new IllegalArgumentException("Expected connectTimeout to be is more that 0.");
+      }
+      try {
```
[the remark this answers is not in the document]
Comment (aivinog1) [reply clients/java/src/main/java/io/zeebe/client/impl/oauth/OAuthCredentialsProviderBuilder.java:224]: Agree, that makes sense, I'll fix that :)

File: clients/java/src/main/java/io/zeebe/client/impl/oauth/OAuthCredentialsProviderBuilder.java [hunk 2]
```diff
@@ -146,9 +193,17 @@ private void applyDefaults() {
...
+      connectTimeout = DEFAULT_CONNECT_TIMEOUT;
+    }
+
+    if (readTimeout == null) {
+      readTimeout = DEFAULT_READ_TIMEOUT;
+    }
   }
 
-  private void validate() {
+  private void validate() throws IllegalArgumentException {
```
[the remark this answers is not in the document]
Comment (aivinog1) [reply clients/java/src/main/java/io/zeebe/client/impl/oauth/OAuthCredentialsProviderBuilder.java]: Done

File: clients/java/src/test/java/io/zeebe/client/OAuthCredentialsProviderTest.java [hunk 3]
```diff
@@ -442,7 +513,20 @@ private void mockCredentials(final String accessToken) {
     mockCredentials(accessToken, AUDIENCE);
   }
 
+  private void mockCredentials(final Integer readDelay) {
```
[the remark this answers is not in the document]
Comment (aivinog1) [reply clients/java/src/test/java/io/zeebe/client/OAuthCredentialsProviderTest.java]: Done

File: clients/java/src/main/java/io/zeebe/client/impl/oauth/OAuthCredentialsProviderBuilder.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (aivinog1) [reply clients/java/src/main/java/io/zeebe/client/impl/oauth/OAuthCredentialsProviderBuilder.java]: Done. I've also made some refactoring to determine whether the 
timeout value in milliseconds included in the Integer range :) I think, that now it looks prettier :)

File: clients/java/src/main/java/io/zeebe/client/impl/oauth/OAuthCredentialsProviderBuilder.java [hunk 4]
```diff
@@ -163,8 +216,29 @@ private void validate() {
         throw new IllegalArgumentException(
             "Expected specified credentials cache to be a file but found directory instead.");
       }
+      if (connectTimeout.isZero() || connectTimeout.isNegative()) {
+        throw new IllegalArgumentException("Expected connectTimeout to be is more that 0.");
+      }
```
Comment (MiguelPires) [inline clients/java/src/main/java/io/zeebe/client/impl/oauth/OAuthCredentialsProviderBuilder.java:221]:
```suggestion
```
This check is duplicated and can be removed.

File: clients/java/src/main/java/io/zeebe/client/impl/oauth/OAuthCredentialsProviderBuilder.java [hunk 5]
```diff
@@ -163,8 +216,29 @@ private void validate() {
...
+
+  private void validateTimeout(final Duration timeout, final String timeoutName) {
+    final boolean valid;
+    if (timeout.isZero() || timeout.isNegative()) {
+      valid = false;
+    } else {
+      final long millis = timeout.toMillis();
+      valid = millis < Integer.MAX_VALUE && millis > Integer.MIN_VALUE;
+    }
+    if (!valid) {
```
Comment (MiguelPires) [inline clients/java/src/main/java/io/zeebe/client/impl/oauth/OAuthCredentialsProviderBuilder.java:237]:
```suggestion
    if (timeout.isZero() || timeout.isNegative() || timeout.toMillis() > Integer.MAX_VALUE) {
```
The `> Integer.MIN_VALUE` is redundant at this point so we can check this more a bit more concisely

**Review by miguelpires — CHANGES_REQUESTED** [review]
Thanks @aivinog1, I just had two minor comments. When you push, you can already squash the commits and I'll push into the repo so it builds

File: clients/java/src/main/java/io/zeebe/client/impl/oauth/OAuthCredentialsProviderBuilder.java (hunk 4 above)
Comment (aivinog1) [reply clients/java/src/main/java/io/zeebe/client/impl/oauth/OAuthCredentialsProviderBuilder.java]: Done

File: clients/java/src/main/java/io/zeebe/client/impl/oauth/OAuthCredentialsProviderBuilder.java (hunk 5 above)
Comment (aivinog1) [reply clients/java/src/main/java/io/zeebe/client/impl/oauth/OAuthCredentialsProviderBuilder.java]: Done

**Comment by aivinog1** [comment]
@MiguelPires Hi! I've squashed commits, can you trigger the build, please? :)

**Review by miguelpires — APPROVED** [review]
Everything looks good. Thanks @aivinog1! 

bors r+

**Pull request merged**